Sources/NIO/DatagramVectorReadManager.swift

```diff
@@ -0,0 +1,203 @@
+    /// that large buffer up into multiple buffers, prepares a series of AddressedEnvelope structures
+    /// appropriately, and then returns that result to the caller.
+    ///
+    /// - warning: If buffer is not at least 1.5kB times `messageCount`, this will almost certainly lead to
+    ///     dropped data. Caution should be taken to ensure that the RecvByteBufferAllocator is allocating an
+    ///     appropriate amount of memory.
+    ///
+    /// - parameters:
+    ///     - socket: The underlying socket from which to read.
+    ///     - buffer: The single large buffer into which reads will be written.
+    func readFromSocket(socket: Socket, buffer: inout ByteBuffer) throws -> ReadResult {
+        assert(buffer.readerIndex == 0, "Buffer was not cleared between calls to readFromSocket!")
+
+        let messageSize = buffer.capacity / self.messageCount
+
+        let result = try buffer.withVeryUnsafeMutableBytes { bufferPointer -> IOResult<Int> in
+            for i in 0..<self.messageCount {
+                // TODO(cory): almost all of this except for the iovec could be done at allocation time. Maybe we should?
+
+                // First we set up the iovec and save it off.
+                self.ioVector[i] = IOVector(iov_base: bufferPointer.baseAddress! + (i * messageSize), iov_len: messageSize)
+
+                // Next we set up the msghdr structure. This points into the other vectors.
+                let msgHdr = msghdr(msg_name: self.sockaddrVector.baseAddress! + i ,
+                                    msg_namelen: socklen_t(MemoryLayout<sockaddr_storage>.size),
+                                    msg_iov: self.ioVector.baseAddress! + i,
+                                    msg_iovlen: 1,  // This is weird, but each message gets only one array. Duh.
+                                    msg_control: nil,
+                                    msg_controllen: 0,
+                                    msg_flags: 0)
+                self.messageVector[i] = MMsgHdr(msg_hdr: msgHdr, msg_len: 0)
+
+                // Note that we don't set up the sockaddr vector: that's because it needs no initialization,
+                // it's written into by the kernel.
+            }
+
+            // We've set up our pointers, it's time to get going. We now issue the call.
+            return try socket.recvmmsg(msgs: self.messageVector)
+        }
+
+        switch result {
+        case .wouldBlock(let messagesProcessed):
+            assert(messagesProcessed == 0)
+            return .none
+        case .processed(let messagesProcessed):
+            return self.buildMessages(messageCount: messagesProcessed, sliceSize: messageSize, buffer: &buffer)
+        }
+    }
+
+    /// Destroys this vector read manager, rendering it impossible to re-use. Use of the vector read manager after this is called is not possible.
+    mutating func destroy() {
```

Comment:
I don't really like that `init` and `destroy` need to be paired exactly 1:1 without any way of checking that. Could we move the allocations to a function called `allocate` and the destruction into `deallocate`?